src/Text.mo

```diff
@@ -226,6 +226,209 @@ module {
+
+  public func matchOfPattern(pat : Pattern) : (cs : Iter.Iter<Char>) -> Match {
+     switch pat {
+       case (#char p) {
+         func (cs : Iter.Iter<Char>) : Match {
+           switch (cs.next()) {
+             case (?c) { if (p == c) #success else (#fail (singleton(c))) };
+             case null { #fail (empty()) };
+           }
+         }
+       };
+       case (#pred p) {
+         func (cs : Iter.Iter<Char>) : Match {
+           switch (cs.next()) {
+             case (?c) { if (p(c)) #success else (#fail (singleton(c))) };
+             case null { #fail (empty()) };
+           }
+         }
+       };
+       case (#text p) {
+         func (cs : Iter.Iter<Char>) : Match {
+           var i = 0;
+           let ds = p.chars();
+           loop {
+             switch (ds.next()) {
+               case (?d)  {
+                 switch (cs.next()) {
+                   case (?c) {
+                     if (c != d) {
+                       return #fail (add(take(i, p.chars()),c))
+                     };
+                     i += 1;
+                   };
+                   case null {
+                     return #fail (take(i, p.chars()));
+                   }
+                 }
+               };
+               case null { return #success };
+             }
+           }
+         }
+       }
+     }
+  };
+
+  /// Returns the sequence of fields in `t`, derived from left to right, separated by text matching pattern `p`.
+  /// A _field_ is a possibly empty, maximal subtext of `t` not containing a match for `p`.
+  /// A _match_ is any sequence of characters matching the pattern `p`, where
+  /// * `#char c` matches the single character sequence, `c`.
+  /// * `#pred p` matches any single character sequence `c` satisfying predicate `p(c)`.
+  /// * `#text t1` matches multi-character text sequence `t1`.
+  /// Two fields are separated by exactly one match.
+  public func split(t : Text, p : Pattern) : Iter.Iter<Text> {
+    let match = matchOfPattern p;
+    var buff = empty();
+    let chars = t.chars();
+    var cs = object {
+        public func next() : ?Char {
+          switch (buff.next()) {
+            case null (chars.next());
+            case oc oc;
+          }
+        };
+      };
+    var state : { #init; #resume; #done} = #init;
+    var field = "";
+    object {
+      public func next() : ?Text {
+        switch state {
+          case (#done) { return null };
+          case (#init) {
+            loop {
+              switch (match(cs)) {
+                case (#success) {
+                  let r = field;
+                  field := "";
+                  state := #resume;
+                  return ?r
+                };
+                case (#fail cs1) {
+                  switch (cs1.next()) {
+                    case (?c) {
+                      field #= fromChar c;
+                      buff := cs1;
+                    };
+                    case null {
```

Comment:
As we talked about, I think making `Match` contain an explicit variant for "tried to match the empty string" would make this clearer.

Reply:
Done